### PR TITLE
Fix issues with `__host__` and `__device__` definitions

### PR DESCRIPTION
**Pull request by miscco**
We currently only checked whether we are compiling with a cuda compiler, but not whether we are actually compiling in cuda mode.

That meant that certain macros werent properly defined.

### CODE REVIEW

**Review by dkolsen-pgi — CHANGES_REQUESTED** [review]

File: libcudacxx/include/cuda/std/__cccl/compiler.h [hunk 1]
```diff
@@ -58,6 +58,11 @@
 #  define _CCCL_CUDA_COMPILER
 #endif // cuda compiler available
 
+// We need to ensure that we not only compile with a cuda compiler but also compile cuda source files
+#if defined(_CCCL_CUDA_COMPILER) && defined(__CUDACC__)
```
[the remark this answers is not in the document]
Comment (miscco) [reply libcudacxx/include/cuda/std/__cccl/compiler.h:62]: That is correct, but we use this only to check whether we actually need `__host__ __device__` anywhere, which is not the case for nvc++?

[the remark this answers is not in the document]
Comment (dkolsen-pgi) [reply libcudacxx/include/cuda/std/__cccl/compiler.h]: NVC++ in CUDA mode, as indicated by `_NVHPC_CUDA`, needs correct `__host__` and `__device__` annotations in the code.

File: libcudacxx/include/cuda/cmath [hunk 2]
```diff
@@ -31,7 +32,8 @@ _LIBCUDACXX_BEGIN_NAMESPACE_CUDA
 template <class _Tp, _CUDA_VSTD::__enable_if_t<_CCCL_TRAIT(_CUDA_VSTD::is_integral, _Tp), int> = 0>
 _CCCL_NODISCARD _LIBCUDACXX_HIDE_FROM_ABI _CCCL_CONSTEXPR_CXX14 _Tp ceil_div(const _Tp __a, const _Tp __b) noexcept
 {
-  _LIBCUDACXX_DEBUG_ASSERT(__a >= _Tp(0), "cuda::ceil_div: a must be non negative");
+  _LIBCUDACXX_DEBUG_ASSERT(_CCCL_TRAIT(_CUDA_VSTD::is_unsigned, _Tp) || __a >= _Tp(0),
+                           "cuda::ceil_div: a must be non negative");
```
Comment (ericniebler) [inline libcudacxx/include/cuda/cmath:36]: how is this change related to the `__host__`/`__device__` annotations?

**Review by ericniebler — APPROVED** [review]

File: libcudacxx/include/cuda/cmath (hunk 2 above)
Comment (miscco) [reply libcudacxx/include/cuda/cmath:36]: That is most likely a leftover from a different PR that sneaked in

**Review by gonidelis — APPROVED** [review]

File: libcudacxx/include/cuda/std/__cccl/execution_space.h [hunk 3]
```diff
@@ -22,27 +22,18 @@
...
-
-#if defined(_CCCL_CUDA_COMPILER)
+// We need to ensure that we not only compile with a cuda compiler but also compile cuda source files
+#if defined(_CCCL_CUDA_COMPILER) && (defined(__CUDACC__) || defined(_NVHPC_CUDA))
 #  define _CCCL_HOST        __host__
 #  define _CCCL_DEVICE      __device__
 #  define _CCCL_HOST_DEVICE __host__ __device__
 #  define _CCCL_FORCEINLINE __forceinline__
-#else // ^^^ _CCCL_CUDA_COMPILER ^^^ / vvv !_CCCL_CUDA_COMPILER
+#else // ^^^ _CCCL_CUDA_COMPILATION ^^^ / vvv !_CCCL_CUDA_COMPILATION vvv
```
Comment (jrhemstad) [inline libcudacxx/include/cuda/std/__cccl/execution_space.h:31]: nit: It doesn't look like we're actually defining `_CCCL_CUDA_COMPILATION`?
